Subscription management: queue resubscribe to fire after unsubscribe succeeded

TimBroddin opened the pull request:
Closes #78626

## Proposed Changes

This PR changes the behavior of the unsubscribe toast. Clicking Resubscribe will now only fire after the Unsubscribe mutation has succeeded.

`useRef` is used instead of `useState` because the callback function would otherwise capture the state at the moment of invocation.

## Testing Instructions

- Apply this PR
- Go to `http://calypso.localhost:3000/subscriptions/sites`
- Unsubscribe from a site & resubscribe quickly
- Refresh, the site should still be there

To more easily test this, add this before the API call in `packages/data-stores/src/reader/mutations/use-site-unsubscribe-mutation.ts` (line 42):

`await new Promise(r => setTimeout(r, 10000));`

This will introduce a 10s timeout.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

TimBroddin committed:
Resubscribe after unsubscribe succeeded

TimBroddin committed:
Resubscribe after unsubscribe succeeded

atanas-dev committed:
Jetpack Pro Dashboard: Remove the ability to detach licenses (#78676)

Remove the detach button for licenses. License should only be issued, assigned once, and revoked from now on.

markbiek committed:
Switch from the domain_upsell task to the domain_customize task (#78545)

* Switch from the domain_upsell task to the domain_customize task

* Fix bug that breaks domain upsell

Even though domain_customize behaves the same way, domain_upsell still exists and needs to be handled

phcp commented on client/landing/subscriptions/components/site-subscriptions-manager/site-row.tsx:
```diff
@@ -137,7 +140,11 @@ const SiteRow = ( {
+						} else {
+							resubscribe( { blog_id, url, doNotInvalidateSiteSubscriptions: true } );
+						}
 						recordSiteResubscribed( {
```
I noticed the resubscribe event is being tracked for every click, is it expected?
<img width="415" alt="image" src="https://github.com/Automattic/wp-calypso/assets/3113712/366c013f-adb8-4775-9d2e-5e9c161aea30">

phcp reviewed (approved):
Good job @TimBroddin! The PR works as expected. Just left one comment, but if it is an expected behavior feel free to move forward and merge the PR.

arthur791004 committed:
Unify Design Preview: Support the font selection for the designs without style variations (#78313)

* Design Picker: Use desgin-preview regardless of the style variations

* Design Preview: Display fonts if no style variations

* Support mobile layout for Fonts

* Apply font styles to preview and submit to the API

* Check the selected font to for unlocking theme

* Fix types

* Design Preview: Avoid displaying the variaions behind the Continue button

* Design Preview: Avoid displaying font variations when the style variations are still loading

* Remove style heading as the styles are split into different sections

* Fix font name might not exist

* Hide description for the virtual themes

* Fix styles of the sticky action button

* Create useRecipe hook to preserve the selected design and variations

* Don't query Colors & Fonts without selected value

* Fix tests

* Fix PremiumGlobalStylesUpgradeModal doesn't show

* Remove is_virtual search param

* Address feedback

TimBroddin committed:
Subscriber page: Add Tracks events (#78675)

* Add Tracks events

* Add usages

* Add tracking of AddSubscriberForm

millerf committed:
Rotate key (#78686)

Initsogar committed:
BackupBrowser: Add file details scaffolding and row counts for database tables (#78514)

* Map row_count from backup ls API response

* Add new FileInfoCard component

This component will be in charge of rendering the file details depending
on the item type. For now it only works for database tables, displaying
the number of rows.

* Render FileInfoCard when user clicks an item

This includes a small refactor, where any item could be opened, not just
the one that `hasChildren` is true.

* Adjust FileCardInfo styles

* Add translate for rows label

* Display file info only on clicked node

The idea is to track what is the node path currently active, so we could
display the file info and hide it for every other node.

* Do not render file info if item has children

millerf committed:
Revert "Rotate key (#78686)" (#78688)

This reverts commit 73e732446169f0d2ecb727e7bdd17abd61392b8e.

sirbrillig committed:
Checkout: Add div inside FormLabel for ThirdPartyDevsAccount to separate styles (#78687)

millerf committed:
Revert "Revert "Rotate key (#78686)" (#78688)" (#78690)

This reverts commit 8c710bf48b1593659f5d331c2f093c3f88599be6.

millerf committed:
Revert "Rotate key (#78686)" (#78693)

This reverts commit 73e732446169f0d2ecb727e7bdd17abd61392b8e.

Commit:
Rotate e2e store and file secrets (#78696)

markbiek committed:
Top-align launchpad skip button (#78649)

This moves the launchpad skip button to the top of the screen and
changes the copy (except on the Newsletter flow).

noahtallen committed:
React 18 (#77046)

* Update packages to be compatible with React 18

* Update internal peer dependencies

* Resolve peer dependency issues

* Update several internal tsconfig references

* Add stream fallback in webpack

* Attempt removing jsx import alias

* Update snapshots to remove whitespace

* Remove custom @types/react patch and pin react 18 types

* Fix issues with children type in React props

* Fix type issues related to useCallback function arguments

* Fix issues with the store selector types

* Remove some uneeded React types

* remove an unused ts-expect-error call

* Add ReactNode types to an inline function component

* Change a use of a component to call it directly

* Fix a type related to interpolate element (ReactNode->ReactElement)

* Cast known render values to string

* Fix context types in legacy class component

* Fix incorrect prop type (Node->ReactNode)

* Fix a type issue with an action creator

* Remove fallback translation which was causing a type error

* Fix an odd issue where children were passed as a function

* Improve generics for use track callback useCallback

* Cast render type to ReactNode and add comment

* Cast const array to readonlyarray so we can use .includes on it

* Another yarn.lock update

* Enable allowDeclareFields for typescript classes in babel preset

* Fix babel config preset syntax

* Try with casting for now

* Fix a bunch of package unit tests

* Fix a bunch of client unit tests

* Conditionally run layout effects in SSR

* Import react dom server from browser bundle explicitly

* Attempt defining TextEncoder globally for tests

* Attempt downgrading use-subscription to fix SSR bug

---------

Co-authored-by: Marin Atanasov <tyxla@abv.bg>

markbiek committed:
Update launchpad skip E2E tests (#78705)

This accounts for the change in button text

alshakero committed:
Stepper: Bulk domain transfer improvements (#78684)

yansern committed:
Subscription Management: Rework recommended site tracks implementation. (#78601)

* Fix type issues. (+17 squashed commits)
Squashed commits:
[ab9d938155] Just repositioning prop order.
[0551f21ca3] Fix incorrect filtering of undefined props.
[fe374c3701] Remove unused unsubscribed tracks.
[2bb45aa9df] Export tracks event.
[ed909a2e10] Rework recommended sites tracks implementation.
[bf186b7cce] Fix type. Mark it optional.
[3d74702614] Add feed_id and source.
[7e179617ed] Further clean up for resubscribed & unsubscribed tracks event.
[239f5ee8d1] Add reader parity for site subscribed & unsubscribed event.
[24bf0f99e0] Add tracks event calypso_subscriptions_site_resubscribed.
[7a7b305cee] Add helper props isSubscriptionsPortal & isReaderPortal.
[064d3856a6] Add feed_id and source prop to site-icon/title/url-clicked tracks event.
[fa6ba63545] Get rid of source.
[fab7c3d446] Return is not needed.
[fb356211a4] Add tracks event for site subscribed & unsubscribed.
[1cea6fbc61] Comment out tracks for other PRs.
[fe24d4c373] Rework tracks in subscriptions.

* Fix incomplete doc.

* Clean up.

* Clean up.

* Move to separate constants file.

* Improve typings; Include the correct rec_blog_id value with train-tracks events; Use URL value for recommended sites, when blog_url is not present

* Remove mistakenly included file

* Fix up RecommendedSiteType

---------

Co-authored-by: Maciej Grabowski <littlesoldiergraba@gmail.com>

noahtallen committed:
Fix crash related to useEffect returning a value instead of a function (#78708)

n3f committed:
Allow empty strings in memberships price field. (#78704)

creativecoder committed:
Reader: wrap slug/timestamp text to prevent overflow (#78691)

taipeicoder committed:
Make theme type badge focusable (#78475)

taipeicoder committed:
Fix CSS specificity for more button (#78667)

arthur791004 committed:
Assembler: Fix selected layout and styles are not restored after checkout (#78668)

miksansegundo committed:
Assembler - Colors and fonts boxes border is missing on Safari  (#78670)

* Fix border issue on Safari by changing the design only 1px

* Remove empty line

southp committed:
Stepper: fix the signup completion event. (#78333)

* Fire the signup completion event upon all the processing actions being
completed in the processing step.

* Fill in the missing props of the signup completion event as much as
possible

* Do not fire the signup completion event in the launchpad, since it
shouldn't really act as part of the signup flow.

* Add a `toStepper` mechanism to avoid the signup completion event from
firing. It is specifically for Stepper using /start/account to create an
account.

* remove `recordSignupComplete` from Launchpad

* Fix product slug derivation logics and remove `now`

* A vital typo: domainCarItem => domainCartItem

* Fix typing errors.

* Fix type error where the type should be string

---------

Co-authored-by: Dehan <dehandecroos@gmail.com>

worldomonation committed:
Permafail E2E: fix Google Authentication spec. (#78663)

* packages/calypso-e2e/src/lib/pages/login-page.ts
- use accessible selector to locate the Login with Google button.

packages/calypso-e2e/src/lib/pages/external/google-login-page.ts
- update methods to use accessibble locators.

* test/e2e/specs/authentication/authentication__google.ts
- update step names.
- handle double 2FA challenge that Google has been throwing at us, where two 2FA challenges are shown for TOTP.
- codify expected URL paths.

* test/e2e/specs/authentication/authentication__google.ts
- add a boatload of comments explaining the reasoning behind each line.
- fix logic issue with waiting on `page.on` - this should have been a `googlePopupPage.on`.

timur987 committed:
Ensure Credits tab has its right margin (#78533)

* Ensure Credits tab has its right margin

* Replace pixel based solution with overflow:hidden

* Revert the ability to scroll tabs

rjchow committed:
add: wc core profiler flow styling for 2fa (#78674)

* add: wc core profiler flow styling for 2fa

yashwin committed:
Jetpack Pro Dashboard: implement resend SMS verification code for downtime monitoring (#78720)

TimBroddin commented:
Closing this PR. Something went wrong during rebase. Will make a new one.

jsnajdr commented:
> Something went wrong during rebase. Will make a new one.

You could just create a new local `add/subscribers-page-resubscribe-queue` branch, from scratch, totally unrelated to the old one, and force-push it to the GitHub repo. There shouldn't ever be a reason to close the PR. Only the Git branch is in non-recoverable state, the PR is OK.

TimBroddin commented:
> > Something went wrong during rebase. Will make a new one.
> 
> You could just create a new local `add/subscribers-page-resubscribe-queue` branch, from scratch, totally unrelated to the old one, and force-push it to the GitHub repo. There shouldn't ever be a reason to close the PR. Only the Git branch is in non-recoverable state, the PR is OK.

Thanks for the pointer! I'll do this next time!
